src/loader/bgw_launcher.c

```diff
@@ -588,23 +590,26 @@ launcher_handle_message(HTAB *db_htab)
 	return true;
 }
 
+/*
+ * Wrapper around normal postgres `die()` function to give more context on sigterms.
+ * The default, `bgworker_die()`, can't be used due to the fact that it handles signals
+ * synchronously, rather than waiting for a CHECK_FOR_INTERRUPTS().
+ * Using `die()` means that the launcher will be restarted by the postmaster, even when it has
+ * received a SIGTERM, which we decided was the proper behavior. If users want to
+ * disable the launcher, they can set timescaledb.max_background_workers = 0 and
+ * then we will proc_exit(0) before doing anything else.
+ */
 static void
-bgw_sigterm(SIGNAL_ARGS)
+launcher_sigterm(SIGNAL_ARGS)
 {
 	/*
-	 * The launcher should never be in a critical section, but if it is we
-	 * have no choice but to PANIC since it is unsafe to exit.
+	 * do not use a level >= ERROR because we don't want to exit here but
+	 * rather only during CHECK_FOR_INTERRUPTS
 	 */
-	if (CritSectionCount != 0)
-		ereport(PANIC, (errmsg("TimescaleDB background worker launcher terminated while in a critical section")));
-	ereport(LOG, (errmsg("TimescaleDB background worker launcher terminated by administrator command. Launcher will not restart after exiting")));
-	proc_exit(0);
-}
-
-static void
-bgw_sigint(SIGNAL_ARGS)
-{
-	ereport(ERROR, (errmsg("TimescaleDB background worker launcher canceled by administrator command. Launcher will restart after exiting")));
+	ereport(LOG,
```

Comment:
This might cause issues in the bgw tests when we send background worker output to a database in a transaction; is there a chance we'll reenter the signal handler?